Show classic template in the inserter only for specific templates

gigitux opened the pull request:
Fixes #5193 
Fixes #6110

With this PR the `Classic Template Block` will be visible in the inserter. But, the block will be visible in the inserter only for the templates:
- Product Search Results
- Products by Category
- Product Catalog
- Products by Tag
- Products by Tag


### Testing


#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

Check out this branch

1. Open the FSE editor.
2. Be sure that the `Classic Template Block` is visible in the inserter only for the templates specificated above.
3. During the tests, be sure that the name in the title and the name in the description correspond to the template that you are editing.
4. During the tests, edit/save templates too, to be sure that this PR doesn't break the current experience.

![image](https://user-images.githubusercontent.com/4463174/175335852-2d3f7bf7-dbf4-40d3-9747-d2e86f9029fa.png)



* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

### Changelog

> N/A

gigitux committed:
Show classic template in the inserter only for specific templates

gigitux committed:
Merge branch 'trunk' into fix/5193-classic-template-only-for-specific-template

gigitux committed:
add feature gate

gigitux committed:
add todo comment

dinhtungdu reviewed (commented):
@gigitux This is working as expected for built-in templates from the plugin but not for templates from the theme. I tested overriding the Product Catalog template in theme and got this issue:

<img width="1616" alt="image" src="https://user-images.githubusercontent.com/5423135/175753937-8008ff6f-2c68-48c3-9ebc-e24f047b9fea.png">

dinhtungdu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -99,51 +105,128 @@ const Edit = ( { clientId, attributes }: Props ) => {
+// @todo Refactor when there will be possible to show a block according on a template/post with a Gutenberg API. https://github.com/WordPress/gutenberg/pull/41718
+
+if ( isExperimentalBuild() ) {
+	subscribe( () => {
```
Looks like this is being called too often. Do you think that we should debounce it?

gigitux replied:
I don't think that we have space for improvements here :'( 

`subscribe` is triggered every time that an action is dispatched ([docs](https://redux.js.org/api/store#subscribelistener)). The check that I did [here](https://github.com/woocommerce/woocommerce-blocks/blob/6c227af2eec957a785659bd37412ba7b016523f5/assets/js/blocks/classic-template/index.tsx/#L191-L193), it useful for not doing not necessary computation. Do you have any particular ideas?


 BTW, this is should be a workaround until this [PR](https://github.com/WordPress/gutenberg/pull/41718) (or the feature) will land on Gutenberg. If you read the description of the PR that I linked, you can see that for now, [this is the only way to go.](https://github.com/WordPress/gutenberg/issues/41062)

dinhtungdu replied:
What do you think about limiting the Classic Template Block to the Site Editor only?

gigitux replied:
> What do you think about limiting the Classic Template Block to the Site Editor only?

Theoretically, this check is done [here](https://github.com/woocommerce/woocommerce-blocks/pull/6539/files/6c227af2eec957a785659bd37412ba7b016523f5..1ac90a3b68cc5dd17a9b4e35260db8e4d01f1512#diff-49cf8439043f78ee9f68b6374ef821be0b59c5804fec85e9aecf528d7fe9bcecR197-R198). If your idea is always on frontend side, I'm pretty sure that we should use always the `subscribe` function because we need to read the value in the state at every update.

dinhtungdu replied:
@gigitux I'm sorry I wasn't very clear! I mean only controlling the block loading on the PHP side. Like what we did (and removed) with the Mini Cart block, see https://github.com/woocommerce/woocommerce-blocks/pull/6351/files.

dinhtungdu replied:
But think about the reason for https://github.com/woocommerce/woocommerce-blocks/pull/6351, we also may need to do the same if, in the future, we allow editing the 'Product Catalog' template from the Shop page edit screen.

dinhtungdu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -99,51 +105,128 @@ const Edit = ( { clientId, attributes }: Props ) => {
-	example: {
-		attributes: {
-			isPreview: true,
+let currentValue: string | undefined;
```
The name of this variable is not very clear about the data it holds at this point. From the source, I guess it is the current editing template. So I think we should make it more explicit by changing the variable name to something like `currentTemplateId`.

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/5193-classic-template-only-for-specific-template

gigitux committed:
rename variable

gigitux commented:
> @gigitux This is working as expected for built-in templates from the plugin but not for templates from the theme. I tested overriding the Product Catalog template in theme and got this issue:
> 
> <img alt="image" width="1616" src="https://user-images.githubusercontent.com/5423135/175753937-8008ff6f-2c68-48c3-9ebc-e24f047b9fea.png">

Thanks for the review and for testing! Great catch! 💪

I addressed the bug, the problem was [the hardcoded slug](https://github.com/woocommerce/woocommerce-blocks/pull/6539/commits/2ab7ef735dc3636693af4c28b37c592c7028740b#diff-49cf8439043f78ee9f68b6374ef821be0b59c5804fec85e9aecf528d7fe9bcecR121). Now, it should be okay!

dinhtungdu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -99,51 +105,124 @@ const Edit = ( { clientId, attributes }: Props ) => {
+
+if ( isExperimentalBuild() ) {
+	subscribe( () => {
+		const previousValue = currentTemplateId;
```
Should we change this to `previousTemplateId` for consistency?

dinhtungdu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -99,51 +105,124 @@ const Edit = ( { clientId, attributes }: Props ) => {
+			html: false,
+			multiple: false,
+			reusable: false,
+			inserter: true,
```
This should be feature gated, otherwise, the block will be always available in the inserter when the block phase is less than 3.

dinhtungdu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -99,51 +105,124 @@ const Edit = ( { clientId, attributes }: Props ) => {
+		}
+
+		if (
+			getBlockType( BLOCK_SLUG ) === undefined &&
```
should we use the `block` variable here?

dinhtungdu reviewed (commented):
@gigitux thanks for the update! I can confirm this PR is working with overridden templates from themes. I left some minor comments, can you please check?

gigitux committed:
address feedback

gigitux commented:
Thanks for the review, @dinhtungdu. With [1ac90a3](https://github.com/woocommerce/woocommerce-blocks/pull/6539/commits/1ac90a3b68cc5dd17a9b4e35260db8e4d01f1512) I addressed feedback. I reply to you about your question. Maybe are you thinking about a PHP approach?

dinhtungdu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -99,51 +105,135 @@ const Edit = ( { clientId, attributes }: Props ) => {
+			 * See https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5861 for more context
+			 */
+			registerClassicTemplateBlock( {
+				template: parsedTemplate ?? 'WooCommerce Classic Template',
```
Because of the condition in line 219, Isn't the `parsedTemplate` variable always truly at this point?

By the way, 'WooCommerce Classic Template' isn't the correct value for the `template` variable, it should be the template ID.

gigitux committed:
not register block on post and page editor

dinhtungdu reviewed (approved):
Thanks so much for the multiple rounds updates! This is LGTM! :shipit:

dinhtungdu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -10,31 +16,43 @@ import { useBlockProps } from '@wordpress/block-editor';
 import './style.scss';
-import { TEMPLATES } from './constants';
+import { BLOCK_SLUG, TEMPLATES } from './constants';
 
 interface Props {
 	attributes: {
 		template: string;
 		align: string;
 	};
+	setAttributes: ( attributes: Record< string, string > ) => void;
 	clientId: string;
 }
 
-const Edit = ( { clientId, attributes }: Props ) => {
+const Edit = ( { clientId, attributes, setAttributes }: Props ) => {
```
This is optional but we can use the `BlockEditProp` type here, for example:

https://github.com/woocommerce/woocommerce-blocks/blob/f2ece167c36b56a8ec26f15fe573dbc22983a654/assets/js/blocks/price-filter/edit.tsx#L31-L34

dinhtungdu commented on assets/js/blocks/classic-template/index.tsx:
```diff
@@ -99,51 +117,142 @@ const Edit = ( { clientId, attributes }: Props ) => {
+				default: 'wide',
+			},
+		},
+		edit: ( { attributes, clientId, setAttributes } ) => {
+			const newTemplate = template ?? ( attributes.template as string );
+
```
Specifying the type here can save a type casting call.
```suggestion
		edit: ( { attributes, clientId, setAttributes }: Props ) => {
			const newTemplate = template ?? attributes.template;

```

gigitux committed:
avoid casting

gigitux commented:
> Thanks so much for the multiple rounds updates! This is LGTM! :shipit:

Thanks for the multiple reviews! 💪
